diesel/src/connection/transaction_manager.rs

```diff
@@ -31,78 +33,222 @@ pub trait TransactionManager<Conn: Connection> {
+    update: ManagerUpdate,
+}
+
+enum ManagerUpdate {
+    UpdateTransaction(TransactionUpdate),
+    MarkManagerAsInError,
+}
+
+struct TransactionUpdate {
+    update_condition: UpdateCondition,
+    transaction_depth_change: TransactionDepthChange,
+}
+
+#[derive(Clone, Copy)]
+enum TransactionDepthChange {
+    IncreaseDepth,
+    DecreaseDepth,
+}
+
+enum UpdateCondition {
+    /// Only update the transaction depth if if the result was Ok
+    AfterResultCheck,
+    // Only update the transaction depth if the content of this variant is OK
+    AfterCheck(QueryResult<()>),
+    /// Always update the result
+    Unconditionally,
+}
+
 impl AnsiTransactionManager {
+    fn run_if_valid<Conn>(
+        conn: &mut Conn,
+        mut action: impl FnMut(&mut Conn, i32) -> Result<OperationOutcome, Error>,
+    ) -> QueryResult<()>
+    where
+        Conn: Connection<TransactionManager = Self> + CommitErrorProcessor,
+    {
+        let transaction_depth = match &mut conn.transaction_state().status {
+            TransactionManagerStatus::InError => return Err(Error::BrokenTransaction),
+            TransactionManagerStatus::Valid(valid_status) => valid_status.transaction_depth(),
+        };
+        let OperationOutcome { result, update } = action(conn, transaction_depth)?;
+        let status = &mut conn.transaction_state().status;
+        match status {
+            TransactionManagerStatus::InError => Err(Error::BrokenTransaction), // Unlikely, but the action actually updated the transaction state and broke it
+            TransactionManagerStatus::Valid(valid_status) => match update {
+                ManagerUpdate::UpdateTransaction(transaction_update) => {
+                    let transaction_delta = match transaction_update.transaction_depth_change {
+                        TransactionDepthChange::IncreaseDepth => 1,
+                        TransactionDepthChange::DecreaseDepth => -1,
+                    };
+                    match transaction_update.update_condition {
+                        UpdateCondition::AfterResultCheck => {
+                            valid_status.change_transaction_depth(transaction_delta, result)
+                        }
+                        UpdateCondition::AfterCheck(result_to_check) => {
+                            let _r = valid_status
+                                .change_transaction_depth(transaction_delta, result_to_check);
+                            result
+                        }
+                        UpdateCondition::Unconditionally => {
+                            let _r =
+                                valid_status.change_transaction_depth(transaction_delta, Ok(()));
+                            result
+                        }
+                    }
+                }
+                ManagerUpdate::MarkManagerAsInError => {
+                    *status = TransactionManagerStatus::InError;
+                    result
+                }
+            },
+        }
```

Comment:
I'm happy making `change_transaction_depth` take `TransactionDepthChange` instead of `i32` (that's also safer). I'm less sold on avoid the nested match. IMO, having the same variant on multiple lines is best avoided (unless you use `if` in your match condition), as this obfuscates the fact that these three branches are actually the same case. I have however shortened the code via destructuring. I will implement the des-nestification if you think it's best, but please have a look at the new version first.